small/matras.c

```diff
@@ -335,6 +337,40 @@ matras_dealloc_range(struct matras *m, matras_id_t range_count)
 	matras_dealloc(m);
 }
 
+/**
+ * Reserve the amount of extents to perform the specified number of touches
+ * using the given `reserve_func` (if any). On success, the following `count`
+ * touches are guaranteed to succeed given the memory source implements the
+ * `reserve_func` appropriately and is not used outside the matras between
+ * the reservation and touches.
+ *
+ * @retval 0 on success, -1 on failure.
+ */
+int
+matras_reserve_touches(struct matras *m, unsigned touch_count)
+{
+	assert(m->reserve_func != NULL);
+
+	/* No `matras_touch` calls planned. */
+	if (touch_count == 0)
+		return 0;
+
+	/* No view opened, the `matras_touch` will be a no-op. */
+	if (!m->head.prev_view)
+		return 0;
+
+	/* The preious view is empty, the `matras_touch` will be a no-op. */
+	if (m->head.prev_view->block_count == 0)
+		return 0;
+
+	/* The root is only copied once - on the first touch. */
+	bool root_is_to_be_copied = m->head.root == m->head.prev_view->root;
+
+	/* We have extents on 2 levels to possibly be copied after root. */
+	unsigned max_extents_copied = touch_count * 2 + root_is_to_be_copied;
+	return m->reserve_func(m->alloc_ctx, max_extents_copied);
```

Comment:
I don't understand the point of the callback. Why not simply return the number of extents that needs to be reserved?

Reply:
That's an option, but it will force the caller site to hold the allocator context and the reservation function for itself.

Seems more native to have it handled right in the matras.

Reply:
Having two allocation functions (`reserve` and `alloc) complicates the API for users that don't need to reserve blocks.

> Seems more native to have it handled right in the matras.

It isn't handled in matras, actually. Reservations are performed by the callback while matras just does the math. It is the callback that is responsible for storing the reserved blocks somewhere and handing them over to future `alloc` calls.

If matras itself reserved blocks with `alloc`, stashed them somewhere internally, and used them instead of `alloc`, that would be a different story, but I'm not sure this is what you need.

Reply:
Fair enough, the function indeed does nothing except passing the compputed value to the given callback.

```diff
--- a/include/small/matras.h
+++ b/include/small/matras.h
@@ -140,7 +140,6 @@ typedef uint32_t matras_id_t;
  */
 typedef void *(*matras_alloc_func)(void *ctx);
 typedef void (*matras_free_func)(void *ctx, void *ptr);
-typedef int (*matras_reserve_func)(void *ctx, unsigned count);
 
 /** Matras statistics. */
 struct matras_stats {
@@ -188,8 +187,6 @@ struct matras {
        matras_alloc_func alloc_func;
        /* External extent deallocator */
        matras_free_func free_func;
-       /* External extent reserver */
-       matras_reserve_func reserve_func;
        /* Argument passed to extent allocator */
        void *alloc_ctx;
        /* Matras statistics. Can't be NULL (points to dummy if unset). */
@@ -237,15 +234,11 @@ matras_stats_create(struct matras_stats *stats)
  * If stats is NULL, a dummy thread-local variable will be
  * used instead. This means that in this case the matras
  * may be used only in the thread that created it.
- *
- * The reserve_func is optional (can be NULL). In that case
- * the `matras_reserve` will not be functional (asserted).
  */
 void
 matras_create(struct matras *m, matras_id_t extent_size, matras_id_t block_size,
              matras_alloc_func alloc_func, matras_free_func free_func,
-             matras_reserve_func reserve_func, void *alloc_ctx,
-             struct matras_stats *stats);
+             void *alloc_ctx, struct matras_stats *stats);
 
 /**
  * Free all memory used by an instance of matras and
@@ -297,16 +290,11 @@ void
 matras_dealloc_range(struct matras *m, matras_id_t range_count);
 
 /**
- * Reserve the amount of extents to perform the specified number of touches
- * using the given `reserve_func` (if any). On success, the following `count`
- * touches are guaranteed to succeed given the memory source implements the
- * `reserve_func` appropriately and is not used outside the matras between
- * the reservation and touches.
- *
- * @retval 0 on success, -1 on failure.
+ * Calculate the amount of extents required to perform the specified number of
+ * touches.
  */
 int
-matras_reserve_touches(struct matras *m, unsigned count);
+matras_max_extents_required_for_touch(struct matras *m, unsigned touch_count);
 
 /**
  * Convert block id into block address.
diff --git a/small/matras.c b/small/matras.c
index 0d01125..9d696d4 100644
--- a/small/matras.c
+++ b/small/matras.c
@@ -44,8 +44,7 @@ matras_log2(matras_id_t val)
 void
 matras_create(struct matras *m, matras_id_t extent_size, matras_id_t block_size,
              matras_alloc_func alloc_func, matras_free_func free_func,
-             matras_reserve_func reserve_func,  void *alloc_ctx,
-             struct matras_stats *stats)
+             void *alloc_ctx, struct matras_stats *stats)
 {
        /*extent_size must be power of 2 */
        assert((extent_size & (extent_size - 1)) == 0);
@@ -64,7 +63,6 @@ matras_create(struct matras *m, matras_id_t extent_size, matras_id_t block_size,
        m->extent_count = 0;
        m->alloc_func = alloc_func;
        m->free_func = free_func;
-       m->reserve_func = reserve_func;
        m->alloc_ctx = alloc_ctx;
        m->stats = stats ? stats : &dummy_stats;
 
@@ -338,19 +336,12 @@ matras_dealloc_range(struct matras *m, matras_id_t range_count)
 }
 
 /**
- * Reserve the amount of extents to perform the specified number of touches
- * using the given `reserve_func` (if any). On success, the following `count`
- * touches are guaranteed to succeed given the memory source implements the
- * `reserve_func` appropriately and is not used outside the matras between
- * the reservation and touches.
- *
- * @retval 0 on success, -1 on failure.
+ * Calculate the amount of extents required to perform the specified number of
+ * touches.
  */
 int
-matras_reserve_touches(struct matras *m, unsigned touch_count)
+matras_max_extents_required_for_touch(struct matras *m, unsigned touch_count)
 {
-       assert(m->reserve_func != NULL);
-
        /* No `matras_touch` calls planned. */
        if (touch_count == 0)
                return 0;
@@ -367,8 +358,7 @@ matras_reserve_touches(struct matras *m, unsigned touch_count)
        bool root_is_to_be_copied = m->head.root == m->head.prev_view->root;
 
        /* We have extents on 2 levels to possibly be copied after root. */
-       unsigned max_extents_copied = touch_count * 2 + root_is_to_be_copied;
-       return m->reserve_func(m->alloc_ctx, max_extents_copied);
+       return touch_count * 2 + root_is_to_be_copied;
 }
 
 /**
diff --git a/test/matras.cc b/test/matras.cc
index 7b8a994..7bba5a7 100644
--- a/test/matras.cc
+++ b/test/matras.cc
@@ -74,7 +74,7 @@ void matras_alloc_test()
 
        alloc_err_inj_enabled = false;
        for (unsigned int i = 0; i <= maxCapacity; i++) {
-               matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc, pta_free, NULL, NULL, NULL);
+               matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc, pta_free, NULL, NULL);
                fail_unless(mat.capacity == maxCapacity);
                AllocatedItems.clear();
                for (unsigned int j = 0; j < i; j++) {
@@ -117,7 +117,7 @@ void matras_alloc_test()
        }
 
        for (unsigned int i = 0; i <= maxCapacity; i++) {
-               matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc, pta_free, NULL, NULL, NULL);
+               matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc, pta_free, NULL, NULL);
                for (unsigned int j = 0; j < i; j++) {
                        unsigned int res = 0;
                        (void) matras_alloc(&mat, &res);
@@ -133,7 +133,7 @@ void matras_alloc_test()
 
        alloc_err_inj_enabled = true;
        for (unsigned int i = 0; i <= maxCapacity; i++) {
-               matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc, pta_free, NULL, NULL, NULL);
+               matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc, pta_free, NULL, NULL);
 
                alloc_err_inj_countdown = i;
 
@@ -200,7 +200,7 @@ matras_vers_test()
        int cur_num_or_ver = 1;
        struct matras local;
        long extents_in_use = 0;
-       matras_create(&local, VER_EXTENT_SIZE, sizeof(type_t), all, dea, NULL, &extents_in_use, NULL);
+       matras_create(&local, VER_EXTENT_SIZE, sizeof(type_t), all, dea, &extents_in_use, NULL);
        type_t val = 0;
        for (int s = 10; s < 8000; s = int(s * 1.5)) {
                for (int k = 0; k < 800; k++) {
@@ -278,7 +278,7 @@ matras_gh_1145_test()
 
        struct matras local;
        long extents_in_use = 0;
-       matras_create(&local, VER_EXTENT_SIZE, sizeof(type_t), all, dea, NULL, &extents_in_use, NULL);
+       matras_create(&local, VER_EXTENT_SIZE, sizeof(type_t), all, dea, &extents_in_use, NULL);
        struct matras_view view;
        matras_create_read_view(&local, &view);
        matras_id_t id;
@@ -305,7 +305,7 @@ matras_stats_test()
 
        long extents_in_use = 0;
        struct matras matras;
-       matras_create(&matras, VER_EXTENT_SIZE, sizeof(type_t), all, dea, NULL, &extents_in_use, &stats);
+       matras_create(&matras, VER_EXTENT_SIZE, sizeof(type_t), all, dea, &extents_in_use, &stats);
        ok(stats.extent_count == 0);
        ok(stats.read_view_extent_count == 0);
 
@@ -345,7 +345,7 @@ matras_alloc_overflow_test()
        matras_id_t id;
        struct matras mat;
        matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc,
-                     pta_free, NULL, NULL, NULL);
+                     pta_free, NULL, NULL);
        for (size_t i = 0; i < max_capacity; i++) {
                void *data = matras_alloc(&mat, &id);
                fail_unless(data != NULL);
@@ -374,7 +374,7 @@ matras_alloc_range_overflow_test()
        matras_id_t id;
        struct matras mat;
        matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc,
-                     pta_free, NULL, NULL, NULL);
+                     pta_free, NULL, NULL);
        for (size_t i = 0; i < max_capacity; i += range_count) {
                void *data = matras_alloc_range(&mat, &id, range_count);
                fail_unless(data != NULL);
@@ -404,9 +404,8 @@ matras_alloc_reserve_test()
        matras_id_t id;
        struct matras mat;
        int reserved = 0;
-       matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE,
-                     pta_alloc, pta_free, pta_reserve,
-                     static_cast<void *>(&reserved), NULL);
+       matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc,
+                     pta_free, static_cast<void *>(&reserved), NULL);
 
        /* Create an empty matras view. */
        struct matras_view empty_view;
@@ -421,34 +420,25 @@ matras_alloc_reserve_test()
        is(matras_alloc(&mat, &id), NULL);
        is(id, max_capacity - 1);
 
-       /* Reserve with an empty view is a no-op. */
-       reserved = 0;
-       matras_reserve_touches(&mat, 0);
-       is(reserved, 0);
-       matras_reserve_touches(&mat, 1);
-       is(reserved, 0);
-       matras_reserve_touches(&mat, max_capacity / 2);
-       is(reserved, 0);
-       matras_reserve_touches(&mat, max_capacity);
-       is(reserved, 0);
-
-       /* Reserve with no read view is a no-op. */
+       /* Nothing to reserve with an empty view. */
+       is(matras_max_extents_required_for_touch(&mat, 0), 0);
+       is(matras_max_extents_required_for_touch(&mat, 1), 0);
+       is(matras_max_extents_required_for_touch(&mat, max_capacity / 2), 0);
+       is(matras_max_extents_required_for_touch(&mat, max_capacity), 0);
+
+       /* Nothing to reserve with no view. */
        matras_destroy_read_view(&mat, &empty_view);
-       matras_reserve_touches(&mat, 0);
-       is(reserved, 0);
-       matras_reserve_touches(&mat, 1);
-       is(reserved, 0);
-       matras_reserve_touches(&mat, max_capacity / 2);
-       is(reserved, 0);
-       matras_reserve_touches(&mat, max_capacity);
-       is(reserved, 0);
+       is(matras_max_extents_required_for_touch(&mat, 0), 0);
+       is(matras_max_extents_required_for_touch(&mat, 1), 0);
+       is(matras_max_extents_required_for_touch(&mat, max_capacity / 2), 0);
+       is(matras_max_extents_required_for_touch(&mat, max_capacity), 0);
 
        /* Create a filled read view. */
        struct matras_view view;
        matras_create_read_view(&mat, &view);
 
        /* Reserve and touch first and last blocks. */
-       matras_reserve_touches(&mat, 2);
+       reserved = matras_max_extents_required_for_touch(&mat, 2);
        is(reserved, 5);
        matras_touch(&mat, 0);
        matras_touch(&mat, id);
@@ -463,7 +453,7 @@ matras_alloc_reserve_test()
         * blocks after the root is copied.
         */
        matras_touch(&mat, id / 2);
-       matras_reserve_touches(&mat, 2);
+       reserved = matras_max_extents_required_for_touch(&mat, 2);
        is(reserved, 4);
        matras_touch(&mat, 0);
        matras_touch(&mat, id);
@@ -471,9 +461,9 @@ matras_alloc_reserve_test()
 
        /* Reserve and touch random blocks. */
        for (int i = 0; i < random_test_count; i++) {
-               reserved = 0; /* Unreserve blocks. */
                int touch_count = rand() % max_capacity;
-               matras_reserve_touches(&mat, touch_count);
+               reserved = matras_max_extents_required_for_touch(&mat,
+                                                                touch_count);
                for (int j = 0; j < touch_count; j++)
                        matras_touch(&mat, rand() % max_capacity);
                fail_unless(reserved >= 0); /* Do not exceed reserved blocks. */
@@ -485,9 +475,9 @@ matras_alloc_reserve_test()
                matras_destroy_read_view(&mat, &view);
                matras_create_read_view(&mat, &view);
 
-               reserved = 0; /* Unreserve blocks. */
                int touch_count = rand() % max_capacity;
-               matras_reserve_touches(&mat, touch_count);
+               reserved = matras_max_extents_required_for_touch(&mat,
+                                                                touch_count);
                for (int j = 0; j < touch_count; j++)
                        matras_touch(&mat, rand() % max_capacity);
                fail_unless(reserved >= 0); /* Do not exceed reserved blocks. */
```

Reply:
Fair enough, the function indeed does nothing except passing the computed value to the given callback.

<details>

<summary>diff</summary>

```diff
--- a/include/small/matras.h
+++ b/include/small/matras.h
@@ -140,7 +140,6 @@ typedef uint32_t matras_id_t;
  */
 typedef void *(*matras_alloc_func)(void *ctx);
 typedef void (*matras_free_func)(void *ctx, void *ptr);
-typedef int (*matras_reserve_func)(void *ctx, unsigned count);
 
 /** Matras statistics. */
 struct matras_stats {
@@ -188,8 +187,6 @@ struct matras {
        matras_alloc_func alloc_func;
        /* External extent deallocator */
        matras_free_func free_func;
-       /* External extent reserver */
-       matras_reserve_func reserve_func;
        /* Argument passed to extent allocator */
        void *alloc_ctx;
        /* Matras statistics. Can't be NULL (points to dummy if unset). */
@@ -237,15 +234,11 @@ matras_stats_create(struct matras_stats *stats)
  * If stats is NULL, a dummy thread-local variable will be
  * used instead. This means that in this case the matras
  * may be used only in the thread that created it.
- *
- * The reserve_func is optional (can be NULL). In that case
- * the `matras_reserve` will not be functional (asserted).
  */
 void
 matras_create(struct matras *m, matras_id_t extent_size, matras_id_t block_size,
              matras_alloc_func alloc_func, matras_free_func free_func,
-             matras_reserve_func reserve_func, void *alloc_ctx,
-             struct matras_stats *stats);
+             void *alloc_ctx, struct matras_stats *stats);
 
 /**
  * Free all memory used by an instance of matras and
@@ -297,16 +290,11 @@ void
 matras_dealloc_range(struct matras *m, matras_id_t range_count);
 
 /**
- * Reserve the amount of extents to perform the specified number of touches
- * using the given `reserve_func` (if any). On success, the following `count`
- * touches are guaranteed to succeed given the memory source implements the
- * `reserve_func` appropriately and is not used outside the matras between
- * the reservation and touches.
- *
- * @retval 0 on success, -1 on failure.
+ * Calculate the amount of extents required to perform the specified number of
+ * touches.
  */
 int
-matras_reserve_touches(struct matras *m, unsigned count);
+matras_max_extents_required_for_touch(struct matras *m, unsigned touch_count);
 
 /**
  * Convert block id into block address.
diff --git a/small/matras.c b/small/matras.c
index 0d01125..9d696d4 100644
--- a/small/matras.c
+++ b/small/matras.c
@@ -44,8 +44,7 @@ matras_log2(matras_id_t val)
 void
 matras_create(struct matras *m, matras_id_t extent_size, matras_id_t block_size,
              matras_alloc_func alloc_func, matras_free_func free_func,
-             matras_reserve_func reserve_func,  void *alloc_ctx,
-             struct matras_stats *stats)
+             void *alloc_ctx, struct matras_stats *stats)
 {
        /*extent_size must be power of 2 */
        assert((extent_size & (extent_size - 1)) == 0);
@@ -64,7 +63,6 @@ matras_create(struct matras *m, matras_id_t extent_size, matras_id_t block_size,
        m->extent_count = 0;
        m->alloc_func = alloc_func;
        m->free_func = free_func;
-       m->reserve_func = reserve_func;
        m->alloc_ctx = alloc_ctx;
        m->stats = stats ? stats : &dummy_stats;
 
@@ -338,19 +336,12 @@ matras_dealloc_range(struct matras *m, matras_id_t range_count)
 }
 
 /**
- * Reserve the amount of extents to perform the specified number of touches
- * using the given `reserve_func` (if any). On success, the following `count`
- * touches are guaranteed to succeed given the memory source implements the
- * `reserve_func` appropriately and is not used outside the matras between
- * the reservation and touches.
- *
- * @retval 0 on success, -1 on failure.
+ * Calculate the amount of extents required to perform the specified number of
+ * touches.
  */
 int
-matras_reserve_touches(struct matras *m, unsigned touch_count)
+matras_max_extents_required_for_touch(struct matras *m, unsigned touch_count)
 {
-       assert(m->reserve_func != NULL);
-
        /* No `matras_touch` calls planned. */
        if (touch_count == 0)
                return 0;
@@ -367,8 +358,7 @@ matras_reserve_touches(struct matras *m, unsigned touch_count)
        bool root_is_to_be_copied = m->head.root == m->head.prev_view->root;
 
        /* We have extents on 2 levels to possibly be copied after root. */
-       unsigned max_extents_copied = touch_count * 2 + root_is_to_be_copied;
-       return m->reserve_func(m->alloc_ctx, max_extents_copied);
+       return touch_count * 2 + root_is_to_be_copied;
 }
 
 /**
diff --git a/test/matras.cc b/test/matras.cc
index 7b8a994..7bba5a7 100644
--- a/test/matras.cc
+++ b/test/matras.cc
@@ -74,7 +74,7 @@ void matras_alloc_test()
 
        alloc_err_inj_enabled = false;
        for (unsigned int i = 0; i <= maxCapacity; i++) {
-               matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc, pta_free, NULL, NULL, NULL);
+               matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc, pta_free, NULL, NULL);
                fail_unless(mat.capacity == maxCapacity);
                AllocatedItems.clear();
                for (unsigned int j = 0; j < i; j++) {
@@ -117,7 +117,7 @@ void matras_alloc_test()
        }
 
        for (unsigned int i = 0; i <= maxCapacity; i++) {
-               matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc, pta_free, NULL, NULL, NULL);
+               matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc, pta_free, NULL, NULL);
                for (unsigned int j = 0; j < i; j++) {
                        unsigned int res = 0;
                        (void) matras_alloc(&mat, &res);
@@ -133,7 +133,7 @@ void matras_alloc_test()
 
        alloc_err_inj_enabled = true;
        for (unsigned int i = 0; i <= maxCapacity; i++) {
-               matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc, pta_free, NULL, NULL, NULL);
+               matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc, pta_free, NULL, NULL);
 
                alloc_err_inj_countdown = i;
 
@@ -200,7 +200,7 @@ matras_vers_test()
        int cur_num_or_ver = 1;
        struct matras local;
        long extents_in_use = 0;
-       matras_create(&local, VER_EXTENT_SIZE, sizeof(type_t), all, dea, NULL, &extents_in_use, NULL);
+       matras_create(&local, VER_EXTENT_SIZE, sizeof(type_t), all, dea, &extents_in_use, NULL);
        type_t val = 0;
        for (int s = 10; s < 8000; s = int(s * 1.5)) {
                for (int k = 0; k < 800; k++) {
@@ -278,7 +278,7 @@ matras_gh_1145_test()
 
        struct matras local;
        long extents_in_use = 0;
-       matras_create(&local, VER_EXTENT_SIZE, sizeof(type_t), all, dea, NULL, &extents_in_use, NULL);
+       matras_create(&local, VER_EXTENT_SIZE, sizeof(type_t), all, dea, &extents_in_use, NULL);
        struct matras_view view;
        matras_create_read_view(&local, &view);
        matras_id_t id;
@@ -305,7 +305,7 @@ matras_stats_test()
 
        long extents_in_use = 0;
        struct matras matras;
-       matras_create(&matras, VER_EXTENT_SIZE, sizeof(type_t), all, dea, NULL, &extents_in_use, &stats);
+       matras_create(&matras, VER_EXTENT_SIZE, sizeof(type_t), all, dea, &extents_in_use, &stats);
        ok(stats.extent_count == 0);
        ok(stats.read_view_extent_count == 0);
 
@@ -345,7 +345,7 @@ matras_alloc_overflow_test()
        matras_id_t id;
        struct matras mat;
        matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc,
-                     pta_free, NULL, NULL, NULL);
+                     pta_free, NULL, NULL);
        for (size_t i = 0; i < max_capacity; i++) {
                void *data = matras_alloc(&mat, &id);
                fail_unless(data != NULL);
@@ -374,7 +374,7 @@ matras_alloc_range_overflow_test()
        matras_id_t id;
        struct matras mat;
        matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc,
-                     pta_free, NULL, NULL, NULL);
+                     pta_free, NULL, NULL);
        for (size_t i = 0; i < max_capacity; i += range_count) {
                void *data = matras_alloc_range(&mat, &id, range_count);
                fail_unless(data != NULL);
@@ -404,9 +404,8 @@ matras_alloc_reserve_test()
        matras_id_t id;
        struct matras mat;
        int reserved = 0;
-       matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE,
-                     pta_alloc, pta_free, pta_reserve,
-                     static_cast<void *>(&reserved), NULL);
+       matras_create(&mat, PROV_EXTENT_SIZE, PROV_BLOCK_SIZE, pta_alloc,
+                     pta_free, static_cast<void *>(&reserved), NULL);
 
        /* Create an empty matras view. */
        struct matras_view empty_view;
@@ -421,34 +420,25 @@ matras_alloc_reserve_test()
        is(matras_alloc(&mat, &id), NULL);
        is(id, max_capacity - 1);
 
-       /* Reserve with an empty view is a no-op. */
-       reserved = 0;
-       matras_reserve_touches(&mat, 0);
-       is(reserved, 0);
-       matras_reserve_touches(&mat, 1);
-       is(reserved, 0);
-       matras_reserve_touches(&mat, max_capacity / 2);
-       is(reserved, 0);
-       matras_reserve_touches(&mat, max_capacity);
-       is(reserved, 0);
-
-       /* Reserve with no read view is a no-op. */
+       /* Nothing to reserve with an empty view. */
+       is(matras_max_extents_required_for_touch(&mat, 0), 0);
+       is(matras_max_extents_required_for_touch(&mat, 1), 0);
+       is(matras_max_extents_required_for_touch(&mat, max_capacity / 2), 0);
+       is(matras_max_extents_required_for_touch(&mat, max_capacity), 0);
+
+       /* Nothing to reserve with no view. */
        matras_destroy_read_view(&mat, &empty_view);
-       matras_reserve_touches(&mat, 0);
-       is(reserved, 0);
-       matras_reserve_touches(&mat, 1);
-       is(reserved, 0);
-       matras_reserve_touches(&mat, max_capacity / 2);
-       is(reserved, 0);
-       matras_reserve_touches(&mat, max_capacity);
-       is(reserved, 0);
+       is(matras_max_extents_required_for_touch(&mat, 0), 0);
+       is(matras_max_extents_required_for_touch(&mat, 1), 0);
+       is(matras_max_extents_required_for_touch(&mat, max_capacity / 2), 0);
+       is(matras_max_extents_required_for_touch(&mat, max_capacity), 0);
 
        /* Create a filled read view. */
        struct matras_view view;
        matras_create_read_view(&mat, &view);
 
        /* Reserve and touch first and last blocks. */
-       matras_reserve_touches(&mat, 2);
+       reserved = matras_max_extents_required_for_touch(&mat, 2);
        is(reserved, 5);
        matras_touch(&mat, 0);
        matras_touch(&mat, id);
@@ -463,7 +453,7 @@ matras_alloc_reserve_test()
         * blocks after the root is copied.
         */
        matras_touch(&mat, id / 2);
-       matras_reserve_touches(&mat, 2);
+       reserved = matras_max_extents_required_for_touch(&mat, 2);
        is(reserved, 4);
        matras_touch(&mat, 0);
        matras_touch(&mat, id);
@@ -471,9 +461,9 @@ matras_alloc_reserve_test()
 
        /* Reserve and touch random blocks. */
        for (int i = 0; i < random_test_count; i++) {
-               reserved = 0; /* Unreserve blocks. */
                int touch_count = rand() % max_capacity;
-               matras_reserve_touches(&mat, touch_count);
+               reserved = matras_max_extents_required_for_touch(&mat,
+                                                                touch_count);
                for (int j = 0; j < touch_count; j++)
                        matras_touch(&mat, rand() % max_capacity);
                fail_unless(reserved >= 0); /* Do not exceed reserved blocks. */
@@ -485,9 +475,9 @@ matras_alloc_reserve_test()
                matras_destroy_read_view(&mat, &view);
                matras_create_read_view(&mat, &view);
 
-               reserved = 0; /* Unreserve blocks. */
                int touch_count = rand() % max_capacity;
-               matras_reserve_touches(&mat, touch_count);
+               reserved = matras_max_extents_required_for_touch(&mat,
+                                                                touch_count);
                for (int j = 0; j < touch_count; j++)
                        matras_touch(&mat, rand() % max_capacity);
                fail_unless(reserved >= 0); /* Do not exceed reserved blocks. */
```

</details>

Reply:
Still does not sound good to me to deal with memory outside the memory provider (which is matras for a BPS tree), but your points are valid, so won't insist here.

I see no nice way to go.

Reply:
After a discussion, agreed that we should:
1. Create an allocator class to be used with matras (`matras_allocator`). The class should implement extent reservation under the hood using the allocation function provided by the user. BPS tree/vector and LIGHT should take an object of this class instead of alloc/free functions and their context.
2. Add functions to reserve extents with matras. It should be possible to reserve an arbitrary number of extents, as well as the number of extents required to do the specified number of touches.
3. Use matras extent reservation in the BPS tree/vector. Remove the code doing extent reservation both from the BPS tree/vector and the memtx engine (the memtx engine will create a single matras allocator that will be used by all indexes).

Reply:
Implemented point 1 and partially point 2. See the tarantool/tarantool#10734 for details on how the functionality is used in the CE.